Add custom auth callback support

VoletiRam opened the pull request:
This PR adds custom authentication callback support for Valkey modules:

Key Features:
- Support multiple custom authentication callbacks via `valkey_module_auth` macro
- Safe Rust wrappers for blocking and non-blocking authentication mechanisms.
- Implement ACL user authentication via safe wrapper of  `ValkeyModule_AuthenticateClientWithACLUser`
- Enable thread-safe authentication state management between auth handlers and reply callbacks by extending set and get private data APIs in blocked.rs
- Support abort client handling to handle some edge cases like resource allocation failures etc.

Implementation includes safe abstractions around Valkey Module C APIs, providing module developers with flexible/safe interfaces for custom authentication workflows while handling all unsafe conversions internally.

VoletiRam commented:
Please take a look at this PR when you get a chance @KarthikSubbarao @dmitrypol 

Thank you.

dmitrypol commented on src/context/mod.rs:
```diff
@@ -511,6 +511,45 @@ impl Context {
 
+    #[cfg(all(any(
+        feature = "min-redis-compatibility-version-7-2",
+        feature = "min-valkey-compatibility-version-8-0"
```
RedisModule_AuthenticateClientWithACLUser is present in Redis7.0.  Can we add `feature = "min-redis-compatibility-version-7-0"`?

VoletiRam commented:
Thank you for taking the look. Applied the rust formatting and also made some changes in the approach regarding Free Priv data callback.

- Keep Free Priv Data callback to be similar to Auth reply callback and removed unnecessary hashmap and typeID related code.
- Ensure we have free priv callback before registering with core.

KarthikSubbarao commented on examples/auth.rs:
```diff
@@ -0,0 +1,294 @@
+};
+
+#[derive(Debug)]
+enum AuthResult {
+    Allow,
+    Deny,
+    Next,
+}
```
Let's add a comment explaining the intention of each Enum variant

VoletiRam replied:
Ack

KarthikSubbarao commented on src/macros.rs:
```diff
@@ -444,3 +451,39 @@ macro_rules! valkey_module {
+                }
+
+                unsafe {
+                    if let Some(register_auth) = $crate::raw::RedisModule_RegisterAuthCallback {
```
1. Lets make this code `expect` that the RedisModule_RegisterAuthCallback API exists.

```suggestion
                    $crate::raw::RedisModule_RegisterAuthCallback.expect("ValkeyModule_RegisterAuthCallback should exist on engines 7.2 and above")($ctx, Some(__do_auth));
```

2. We should be using feature flags here to ensure this code does not work on engine versions before 7.2

KarthikSubbarao commented on examples/auth.rs:
```diff
@@ -0,0 +1,294 @@
+use valkey_module::alloc::ValkeyAlloc;
+use valkey_module::{
+    valkey_module, Context, Status, ValkeyError, ValkeyString, AUTH_HANDLED, AUTH_NOT_HANDLED,
+};
```
There are a lot of functions here where the intention can be explained in a comment to help readers and to help extend this in the future

VoletiRam replied:
Sure, will make this change.

KarthikSubbarao commented on src/context/blocked.rs:
```diff
@@ -1,18 +1,108 @@
+pub type AuthReplyCallback = fn(&Context, ValkeyString, ValkeyString) -> Result<c_int, ValkeyError>;
+type FreePrivDataCallback<T> = fn(&Context, T);
+
+static mut AUTH_REPLY_CALLBACK: Option<AuthReplyCallback> = None;
+static mut FREE_PRIV_CALLBACK: Option<Box<dyn Fn(&Context, *mut c_void) + Send + Sync>> = None;
```
Having the auth callback and free_priv callbacks as a static will not work since we can have several Module callbacks that are in flight and we would want to track each individual one

VoletiRam replied:
Even after using unique extern C wrappers per blocked client, we still have issue calling the corresponding rust callbacks as users provide the callbacks dynamically in case of blocking auth. Unlike Macros, we can't embed the rust callbacks inside wrappers and hence needed to store the callback pointers in static variables. 

I am considering combining the priv data and callbacks together and store them in Core (similar to timer.rs) and fetch them for core in our reply/free wrappers and extract the priv and callback from it and pass around respectively to callback and free priv data.

KarthikSubbarao commented on src/context/blocked.rs:
```diff
@@ -28,8 +118,37 @@ impl Context {
+        free_privdata_callback: Option<FreePrivDataCallback<T>>,
+    ) -> BlockedClient {
+        unsafe {
+            AUTH_REPLY_CALLBACK = Some(auth_reply_callback);
+
+            let free_callback: Option<unsafe extern "C" fn(*mut raw::RedisModuleCtx, *mut c_void)> =
+                if let Some(callback) = free_privdata_callback {
+                    FREE_PRIV_CALLBACK = Some(Box::new(move |ctx: &Context, data: *mut c_void| {
+                        let value = ptr::read(data as *const T);
+                        callback(ctx, value);
+                    }));
+                    Some(raw_free_callback_wrapper::<T>)
+                } else {
+                    FREE_PRIV_CALLBACK = None;
+                    None
+                };
```
To ensure we can handle multiple inflight blocking auth requests, each time block_client_on_auth is called, we should create a new extern C function and provide it to the core with the `ValkeyModule_BlockClientOnAuth` API.

We can have it such that the Rust callback function will be embedded inside the extern C function, so we know it will get called by the core

KarthikSubbarao replied:
> My understanding was auth calls would always be handled by main thread 

Yes, it will always be called by the main thread. The common scenario is multiple clients using authentication at the same time. If Client 1 is blocked, we can have Client 2 trying to perform blocking authentication. So, we would provide an interface that allows multiple blocking authentications while others are already in progress (without overriding each other)

VoletiRam replied:
Makes sense, will provide the solution to support this.

KarthikSubbarao commented on src/context/mod.rs:
```diff
@@ -511,6 +511,45 @@ impl Context {
         }
     }
 
+    #[cfg(all(any(
```
Let's document the two new APIs here

VoletiRam replied:
Where do we usually document this?

KarthikSubbarao replied:
Oh, I just meant a documentation style comment using `/// Add a comment here explaining purpose, input, output`

KarthikSubbarao commented on src/redismodule.rs:
```diff
@@ -125,6 +127,10 @@ impl ValkeyString {
     }
 
+    pub fn into_raw(self) -> *mut raw::RedisModuleString {
+        self.take() // internally calls the private take() method
```
You could instead access `inner` directly. It is public

VoletiRam replied:
Aah, I see. Thank you.

KarthikSubbarao replied:
See if this helps: https://docs.rs/valkey-module/latest/valkey_module/struct.ValkeyString.html#method.create_and_retain

KarthikSubbarao commented on src/redismodule.rs:
```diff
@@ -38,6 +38,8 @@ impl From<ValkeyError> for ValkeyValueResult {
 
 pub const VALKEY_OK: ValkeyValueResult = Ok(ValkeyValue::SimpleStringStatic("OK"));
 pub const TYPE_METHOD_VERSION: u64 = raw::REDISMODULE_TYPE_METHOD_VERSION as u64;
+pub const AUTH_HANDLED: i32 = raw::REDISMODULE_AUTH_HANDLED as i32;
+pub const AUTH_NOT_HANDLED: i32 = raw::REDISMODULE_AUTH_NOT_HANDLED as i32;
```
Question: Is this the correct integer type?

VoletiRam replied:
I see the other hash defines are u64 type and will change these to be consistent with other.

VoletiRam replied:
Actually the expected integer type here is i32. Will leave it as is

dmitrypol commented on examples/auth.rs:
```diff
@@ -0,0 +1,294 @@
+        ctx.log_debug(&format!("Matched {} user credentials", auth_type));
+        match ctx.authenticate_client_with_acl_user(username) {
+            Status::Ok => {
+                ctx.log_debug(&format!("Successfully authenticated {} user", auth_type));
```
can you make all logging statement `ctx.log_notice`?  This way when running Valkey with default logging settings they show up

dmitrypol commented on examples/auth.rs:
```diff
@@ -0,0 +1,294 @@
+impl fmt::Display for AuthResult {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self {
+            AuthResult::Allow => write!(f, "Allow"),
```
can we do this instead?  

```
        match self {
            _ => write!(f, "{:?}", self),
        }

```

mkmkme replied:
You don't even need `match` in that case

dmitrypol commented on examples/auth.rs:
```diff
@@ -0,0 +1,294 @@
+    ctx.log_debug(&format!(
+        "{} auth attempt for user: {}",
+        auth_type,
+        username.to_string()
```
nit:  you don't need `to_string` in log statements for ValkeyString

VoletiRam commented on src/context/mod.rs:
```diff
@@ -511,6 +511,45 @@ impl Context {
         }
     }
 
+    #[cfg(all(any(
```
Sure, I can move the two APIs to new context/auth.rs.

VoletiRam commented on examples/auth.rs:
```diff
@@ -0,0 +1,294 @@
+        if username_str == "abort-test" {
+            // Abort the client directly from the thread
+            blocked_client.abort().unwrap_or_else(|e| {
+                println!("Failed to abort blocked client: {:?}", e);
```
I initially tried logging with ctx.log_notice but it threw error because Context is not thread safe. I guess I can directly use the log_notice here. Will change it.

VoletiRam commented:
I have addressed your comments. Please take a look when you get a chance.

Here is the summary of the changes in this revision:

1. Now, we can handle multiple inflight authentication requests. Merged blocked reply and free callback pointers with priv data pointer and provide to core so that we can scope the callback lifecycle within the blocked client context and get rid of static storage. 
2. Provide unique callback wrappers for each auth callback to avoid name conflicts with multiple modules and potential overrides by linking the module and callback names to extern C function.
3. Add and separate more comprehensive tests.
4. Add documentation to APIs and comments to examples.
5. Created new context/auth.rs file and moved the relevant APIs there.

Thank you so much for the feedback @dmitrypol @KarthikSubbarao @mkmkme

KarthikSubbarao commented on src/context/blocked.rs:
```diff
@@ -1,18 +1,151 @@
-pub struct BlockedClient {
+pub struct BlockedClient<T = ()> {
     pub(crate) inner: *mut raw::RedisModuleBlockedClient,
+    auth_reply_callback: Option<AuthReplyCallback<T>>,
```
By adding this `auth_reply_callback: Option<AuthReplyCallback<T>>,` to this structure, we are coupling the `BlockedClient` structure to Authentication which does not seem right since there are other modes of blocking: 
1. General: BlockClient
2. Keys: BlockClientOnKeys


We could use an enum here which has variant for each type of ReplyCallback.

This will allow extending this interface to support blocking other modes

KarthikSubbarao commented on src/context/blocked.rs:
```diff
@@ -1,18 +1,151 @@
+    let (_, free_cb, data) = *Box::from_raw(
+        privdata
+            as *mut (
+                AuthReplyCallback<T>,
```
Following the suggestion above (https://github.com/valkey-io/valkeymodule-rs/pull/186/files#r2019112792), I would recommend making this free_callback generic IF we are going with this approach

KarthikSubbarao commented on src/context/blocked.rs:
```diff
@@ -1,18 +1,151 @@
+impl<T> Drop for BlockedClient<T> {
     fn drop(&mut self) {
-        unsafe { raw::RedisModule_UnblockClient.unwrap()(self.inner, ptr::null_mut()) };
+        if !self.inner.is_null() {
```
This will leak boxed data for the general blocking case. Can we instead check for private data and only provide a boxed value if private data exists?

Also, we should still call RedisModule_UnblockClient for the generic case. It is just that private data should be null here

KarthikSubbarao commented on src/context/blocked.rs:
```diff
@@ -1,18 +1,151 @@
+    }
+
+    pub fn set_blocked_private_data(&mut self, data: T) {
+        self.private_data = Some(Box::new(data));
```
Question: If someone were to set private data on a Blocked Client (General blocking, not Auth), will they leak memory? We do not currently accept a free callback for the General Blocking case, so the Module cannot free it.

This will be addressed once we extend this structure to accept free_priv callbacks for the General Blocking case as well.

Do you want to include general blocking in a separate PR or this PR?

If we are extending this structure to handle general blocking (that accepts reply and free callbacks) in a separate PR, then we should add a documentation comment (`///`) saying this should only be used for blockonauth until this interface supports general blocking

dmitrypol commented on src/context/auth.rs:
```diff
@@ -0,0 +1,39 @@
+    /// * `Status::Ok` - Authentication successful
+    /// * `Status::Err` - Authentication failed
+    #[cfg(all(any(
+        feature = "min-redis-compatibility-version-7-0",
```
Wha't about `min-redis-compatibility-version-7-2`?  Why not invert the logic to have 
```
    not(any(
        feature = "min-redis-compatibility-version-6-2",
        feature = "min-redis-compatibility-version-6-0"
    ))
```

dmitrypol commented on Cargo.toml:
```diff
@@ -32,6 +32,11 @@ name = "acl"
+[[example]]
+name = "auth"
+crate-type = ["cdylib"]
+required-features = ["min-valkey-compatibility-version-8-0"]
```
I believe this is also compatible with `min-redis-compatibility-version-7-2`

VoletiRam commented:
I have addressed your comments of the previous revision. Please take a look when you get a chance.

Here is the summary of the changes:
- Added enum type to callbacks to make it generic and extensible to future variations of blockClient APIs.
- Updated feature flags
- Added documentation to API wrappers.

KarthikSubbarao commented on src/context/blocked.rs:
```diff
@@ -1,18 +1,177 @@
+    err: *mut *mut raw::RedisModuleString,
+) -> c_int {
+    let context = Context::new(ctx);
+    let ctx_ptr = std::ptr::NonNull::new_unchecked(ctx);
+    let username = ValkeyString::new(Some(ctx_ptr), username);
+    let password = ValkeyString::new(Some(ctx_ptr), password);
```
There should be an API of the `ValkeyString` struct that accepts a `*mut raw::RedisModuleCtx` directly so that you don't have to box this structure

The other potential issue of using `new` is that it calls string_retain_string which increments a ref counter and this causes a leak of ValkeyString in the core:
```
    pub fn new(
        ctx: Option<NonNull<raw::RedisModuleCtx>>,
        inner: *mut raw::RedisModuleString,
    ) -> Self {
        let ctx = ctx.map_or(std::ptr::null_mut(), |v| v.as_ptr());
        raw::string_retain_string(ctx, inner);
        Self { ctx, inner }
    }
```

KarthikSubbarao commented on src/context/blocked.rs:
```diff
@@ -1,18 +1,177 @@
+        return;
+    }
+
+    let (_, free_cb, data) = *Box::from_raw(
```
While the tuple containing 3 elements works, IMO it is slightly more error prone than using a dedicated structure that defines the elements that priv data should contain.

What do you think?

KarthikSubbarao commented on src/context/blocked.rs:
```diff
@@ -1,18 +1,177 @@
+        if !self.inner.is_null() {
+            let callback_data_ptr = self
+                .private_data
+                .take()
+                .map(|private_data| {
+                    let callbacks_and_data = (
+                        self.reply_callback.take(),
+                        self.free_priv_callback.take(),
+                        private_data,
+                    );
+                    Box::into_raw(Box::new(callbacks_and_data)) as *mut c_void
+                })
+                .unwrap_or_else(std::ptr::null_mut);
```
Do we need both `take` and `map`? Doesn't `map` itself consume the `Option` (so we could use `map` instead of both operations)?

If we really do need both, would suggest splitting this into two lines


Also, I don't think we need `unwrap_or_else` since we could use `map_or_else` for this entire transformation

KarthikSubbarao commented on src/context/blocked.rs:
```diff
@@ -28,8 +187,54 @@ impl Context {
+    /// Retrieves private data associated with a blocked client
+    ///
+    /// Wrapper around ValkeyModule_GetBlockedClientPrivateData.
+    pub fn get_blocked_client_privdata(&self) -> *mut c_void {
```
Instead of returning a raw `* c_void` pointer that contains the SDK's internally managed data structure (currently a tuple), it would be better to return `data: T` just like we do in the `set_blocked_private_data` API

KarthikSubbarao commented on src/macros.rs:
```diff
@@ -489,3 +496,67 @@ macro_rules! valkey_module {
+                        err: *mut *mut $crate::raw::RedisModuleString,
+                    ) -> std::os::raw::c_int {
+                        let context = $crate::Context::new(ctx);
+                        let ctx_ptr = unsafe { std::ptr::NonNull::new_unchecked(ctx) };
+                        let username = $crate::ValkeyString::new(Some(ctx_ptr), username);
+                        let password = $crate::ValkeyString::new(Some(ctx_ptr), password);
```
Same comment as https://github.com/valkey-io/valkeymodule-rs/pull/186/files#r2025799069 to ensure we do not leak the username and password SDS

KarthikSubbarao commented on src/macros.rs:
```diff
@@ -489,3 +496,67 @@ macro_rules! valkey_module {
+                        }
+                    }
+
+                    #[cfg(all(any(
+                        feature = "min-redis-compatibility-version-7-2",
+                        feature = "min-valkey-compatibility-version-8-0"
+                    )))]
```
Question: This will just skip the calling the API below if we are on versions before 7.2. Should we propagate an error back to the caller function instead to indicate that `valkey_module_auth` did not work instead of silently failing?

KarthikSubbarao commented on tests/integration.rs:
```diff
@@ -1167,3 +1170,315 @@ fn test_preload() -> Result<()> {
+        ("blockUserDelay", "blockPassDelay"),
+    ];
+
+    setup_acl_users(&mut con, &users)?;
```
Same comment as https://github.com/valkey-io/valkeymodule-rs/pull/186/files#r2038527295

VoletiRam commented:
I have addressed the comments. Please take a look when you get a chance @KarthikSubbarao @dmitrypol

KarthikSubbarao commented on src/context/blocked.rs:
```diff
@@ -1,18 +1,181 @@
+    /// * `data` - The private data to store
+    pub fn set_blocked_private_data(&mut self, data: T) {
+        if self.free_callback.is_none() {
+            panic!("Cannot set private data without a free callback - this would leak memory");
```
We can return an error instead of panicking here - so this function will return a `Result`

dmitrypol replied:
generally I prefer returning Result and have module developer handle it

dmitrypol commented on src/context/blocked.rs:
```diff
@@ -1,18 +1,181 @@
+    ///
+    /// # Arguments
+    /// * `data` - The private data to store
+    pub fn set_blocked_private_data(&mut self, data: T) {
```
`
    pub fn set_blocked_private_data(&mut self, data: T) -> Result<> {
    if self.free_callback.is_none() {
       return Err("Cannot set private data without a free callback - this would leak memory")
   }
}

`

VoletiRam commented on src/context/blocked.rs:
```diff
@@ -1,18 +1,181 @@
+        None => panic!("[auth_reply_wrapper] Reply callback is null; this should not happen!"),
+    };
+
+    let data_ref = user_private_data.data.as_ref().map(|boxed| &**boxed);
```
Unlike timer, we need to retain the data for the free callback. If I take the ownership of data in auth_reply_wrapper using take, I won't have data available in free callback and wouldn't able to invoke free callback.

There are few options. 

1. I can clone the data in auth_reply_callback
```
let data = user_private_data.data.as_ref().map(|boxed| (**boxed).clone());

```

2. Using references like above.(recommended as it doesn't incur additional memory cost).

3. Use refcell to allow borrowing data in reply callback but it is similar to references with additional wrapper overhead

```
  // Borrow the data mutably if it exists
    let data = user_private_data.data.as_ref().map(|refcell| {
        refcell.borrow_mut().as_mut()
    });
    ```
